### PR TITLE
Adding Hinge embedding loss function

### PR DESCRIPTION
**Pull request by ojhalakshya**
Hello,
This loss function in regarding to my previous issue https://github.com/mlpack/mlpack/issues/2200 in which I opened up for help from all guys working hard to contribute. There I mentioned many loss functions which are now being implemented.
I also am trying to implement the Hinge Embedding Loss Function now.
It can be referred from both [pytorch](https://pytorch.org/docs/stable/_modules/torch/nn/modules/loss.html) and [tensorflow](https://www.tensorflow.org/api_docs/python/tf/keras/losses/Hinge).
Hope to contribute more.
Thanks.

### CODE REVIEW

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+#define MLPACK_METHODS_ANN_LOSS_FUNCTION_HINGE_EMBEDDING_LOSS_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The Hinge Embedding loss function is often used to improve
+ * variational auto encoder. This function is the log of hyperbolic
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp]: This needs to be changed.

File: src/mlpack/tests/loss_functions_test.cpp [hunk 2]
```diff
@@ -396,4 +397,41 @@ BOOST_AUTO_TEST_CASE(DiceLossTest)
...
+    BOOST_REQUIRE_CLOSE(el, 0.0, 1e-5);
+  }
+
+  BOOST_REQUIRE_EQUAL(output.n_rows, input.n_rows);
+  BOOST_REQUIRE_EQUAL(output.n_cols, input.n_cols);
+
+  // Test the Forward function. Loss should be 0.879
+  input = arma::mat("0.1 0.2 0.3 0.4 0.5");
+  target = arma::mat("0.11 0.22 0.33 0.44 0.55");
+    loss = module.Forward(std::move(input), std::move(target));
```
Comment (kartikdutt18) [inline src/mlpack/tests/loss_functions_test.cpp]: No indentation required.

File: src/mlpack/tests/loss_functions_test.cpp [hunk 3]
```diff
@@ -396,4 +397,41 @@ BOOST_AUTO_TEST_CASE(DiceLossTest)
...
+ */
+BOOST_AUTO_TEST_CASE(HingeEmbeddingLossTest)
+{
+  arma::mat input, target, output;
+  double loss;
+
+  // Test the Forward Function. Loss should be 0 if input = target.
+  input = arma::ones(10, 1);
+  target = arma::ones(10, 1);
+  loss - module.Forward(std::move(input), std::move(target));
```
Comment (kartikdutt18) [inline src/mlpack/tests/loss_functions_test.cpp]: This should be = (typo).

Comment (kartikdutt18) [reply src/mlpack/tests/loss_functions_test.cpp]: I think this is one of the error. Could you fix the rest if any.

Comment (ojhalakshya) [reply src/mlpack/tests/loss_functions_test.cpp]: Sure.
Working on it.
Thanks.

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+#define MLPACK_METHODS_ANN_LOSS_FUNCTION_HINGE_EMBEDDING_LOSS_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The Hinge Embedding loss function is often used to improve
+
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp]: Looks like we can remove the empty line here.

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+#ifndef MLPACK_METHODS_ANN_LOSS_FUNCTION_HINGE_EMBEDDING_LOSS_HPP
+#define MLPACK_METHODS_ANN_LOSS_FUNCTION_HINGE_EMBEDDING_LOSS_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The Hinge Embedding loss function is often used to improve
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp]: Looks like there is something missing.

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+ */
+ template<
+        typename InputDataType = arma::mat,
+        typename OutputDataType = arma::mat
+>
+class HingeEmbeddingLoss
+{
+ public:
+  /**
+   * Create the HingeEmbedding object with the specified parameters.
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp]: There are no parameters that can be specified, so probably save to remove that part.

Comment (birm) [reply src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp]:
```suggestion
   * Create the HingeEmbedding object.
```

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp (hunk 4 above)
Comment (birm) [reply src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp]:
```suggestion
```

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp (hunk 5 above)
Comment (birm) [reply src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp]: I'm more interested in what you meant to write, but in case you wanted a suggestion anyway:
```suggestion
 * The Hinge Embedding loss function is often used to improve performance
 * in semi-supervised learning or to learn nonlinear embeddings.
```

File: src/mlpack/tests/loss_functions_test.cpp (hunk 2 above)
Comment (ojhalakshya) [reply src/mlpack/tests/loss_functions_test.cpp]: Thanks for noticing.

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp (hunk 5 above)
Comment (ojhalakshya) [reply src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp]: yeah thanks. A great help!

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp (hunk 4 above)
Comment (ojhalakshya) [reply src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp]: I have made some changes please have a look when you get time.
Thanks.

File: src/mlpack/tests/loss_functions_test.cpp [hunk 7]
```diff
@@ -436,4 +437,42 @@ BOOST_AUTO_TEST_CASE(SimpleMeanBiasErrorTest)
...
+  for (double el : output)
+  {
+    // For input = target we should get 0.0 everywhere.
+    BOOST_REQUIRE_CLOSE(el, 0.0, 1e-5);
+  }
+
+  BOOST_REQUIRE_EQUAL(output.n_rows, input.n_rows);
+  BOOST_REQUIRE_EQUAL(output.n_cols, input.n_cols);
+
+  // Test the Forward function. Loss should be 0.879
```
Comment (kartikdutt18) [inline src/mlpack/tests/loss_functions_test.cpp]: Could you also add a stop here. Thanks for all your work.

Comment (ojhalakshya) [reply src/mlpack/tests/loss_functions_test.cpp]: Taken care of in this commit.
Thanks for noticing.

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType, typename OutputType>
+void HingeEmbeddingLoss<InputDataType, OutputDataType>::Backward(
+    const InputType&& input,
+    const TargetType&& target,
+    OutputType&& output)
+{
+  for (size_t i = 0; i < input.n_elem; i++)
```
Comment (birm) [inline src/mlpack/methods/ann/loss_functions/hinge_embedding_loss_impl.hpp]: I wonder if we can avoid the for loop here too?

**Comment by rcurtin** [comment]
This code touches the ANN code, and there was recently a big refactoring of the ANN code in #2259, so be sure to merge the master branch into your branch here to make sure that nothing will fail if this PR is merged. :+1: (I'm pasting this message into all possibly relevant PRs.)

**Comment by ojhalakshya** [comment]
Hey @zoq @rcurtin 
I am not able to get a hold of this fail, been working to fix this from like 5 days but still no clue.
Apart from this error I have pretty much everything covered up.
Can you help me a bit in this?
The error occurs in the loss_function_test.cpp line no 522 and it generates this error when I use 

`HingeEmbeddingLoss<> module;`

and when I just remove the `<>` 

`HingeEmbeddingLoss module;` 

then the template error doesn't show and only the last 4 errors show up.

I am not sure with the last 4 errors in both the cases as Forward and Backward function are defined correctly and when I build without the loss_function_test.cpp file the build works 100%.

Screenshot from 2020-03-13 10-21-18

Thanks.

**Comment by rcurtin** [comment]
You can see in the error message that it is referencing `struct LossFunctionsTest::HingeEmbeddingLoss`.  Do you have a class or struct called `HingeEmbeddingLoss` defined locally in the `loss_functions_test.cpp` file?  (This might happen if, for instance, you've named one of the tests `HingeEmbeddingLoss`.)

**Comment by ojhalakshya** [comment]
Hi @rcurtin 
Thanks for the reply and the solution also, I was able to fix the same because of it.
Can you please also have a look at the whole pr so if it is fine.
Thanks.

File: HISTORY.md [hunk 9]
```diff
@@ -52,6 +52,8 @@
   * Change neural network types to avoid unnecessary use of rvalue references
     (#2259).
 
+  * Add Hinge-Embedding-Loss Function (#2229).
```
Comment (favre49) [inline HISTORY.md]: I don't think there needs to be hyphens here

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The Hinge Embedding loss function is often used to compute the loss
+ * between y_true and y_pred.
+ *
```
Comment (favre49) [inline src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp]: I think we can remove this newline.

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  //! Locally-stored output parameter object.
+  OutputDataType outputParameter;
+
+  //! Hyperparameter a for smoothening function curve.
+  double a;
```
Comment (favre49) [inline src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp]: I'm not sure what this is for?

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss_impl.hpp [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+template<typename InputDataType, typename OutputDataType>
+HingeEmbeddingLoss<InputDataType, OutputDataType>::HingeEmbeddingLoss()
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType>
+double HingeEmbeddingLoss<InputDataType, OutputDataType>::Forward(
+    const InputType&& input, const TargetType&& target)
```
Comment (favre49) [inline src/mlpack/methods/ann/loss_functions/hinge_embedding_loss_impl.hpp]: I think you missed out the recent change to method signatures in #2259 . You'll have to change your code to conform to it.

File: HISTORY.md (hunk 9 above)
Comment (ojhalakshya) [reply HISTORY.md]: Thanks for noticing, I have changed it in the new commit Kindly have a look.
Thanks.

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp (hunk 10 above)
Comment (ojhalakshya) [reply src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp]: Adjusted in the new commit.
Thanks.

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss_impl.hpp (hunk 12 above)
Comment (ojhalakshya) [reply src/mlpack/methods/ann/loss_functions/hinge_embedding_loss_impl.hpp]: Hi @favre49 
Thanks for noticing.
I have changed these in the new commit, kindly have a look when you get time.
Thanks.

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss_impl.hpp (hunk 8 above)
Comment (ojhalakshya) [reply src/mlpack/methods/ann/loss_functions/hinge_embedding_loss_impl.hpp]: I have implemented the changes, kindly have a look when you get time.
Thanks.

**Comment by birm** [comment]
Azure isn't triggering, so I'll have to check out and build/test manually. See you in about an hour!

**Comment by kartikdutt18** [comment]
Hey @ojhalakshya, One thing I noticed that you need to add the names of functionalities / layers you add to History.md . Thanks.

**Comment by ojhalakshya** [comment]
> When running the test locally, I got the following error:
> 
> ```
> /Users/ryanbirmingham/Desktop/git/mlpack/src/mlpack/tests/loss_functions_test.cpp:545: fatal error: in "LossFunctionsTest/HingeEmbeddingLossTest": difference{5.66667} between loss{4.7999999999999998} and 0.71999997{0.71999997000000004} exceeds 0.001%
> 
> *** 1 failure is detected in the test module "mlpackTest"
> ```

Working on it, Thanks for noticing.
Will ping you when ready.
Thanks.

**Comment by ojhalakshya** [comment]
> Hey @ojhalakshya, One thing I noticed that you need to add the names of functionalities / layers you add to History.md . Thanks.

Are talking about hardshrink and softshrink?

**Comment by kartikdutt18** [comment]
Yes I am, You will also need to do the same for this layer / function as well. Thanks for all the work though 👍 .

**Comment by ojhalakshya** [comment]
Hey @birm 
I have made some changes, I think now everything is fine.
Can you see into this when you get time.
Thanks.

**Comment by ojhalakshya** [comment]
@rcurtin 
Can you kindly have a look when you get time. Although azure pipelines are not working, I did build this atleast 4 times with the tests also and I think it works fine now.
I wanted to know if it works fine in other systems also.
@kartikdutt18 can you also try this one if comfortable.
Thanks.

**Comment by birm** [comment]
I think that azure is working again; can you trick github into running those tests?

**Comment by ojhalakshya** [comment]
I am not sure about the LoadSaveTest in ctest in which the macOS Plain failed, as I can see all builds and all the corresponding tests had passed successfully see [here](https://dev.azure.com/mlpack/mlpack/_build/results?buildId=1015&view=logs&j=1b38ea94-fede-5667-20eb-44d3de1ec085&t=72087b03-c17c-5a26-b713-236d9bb6c096).
@kartikdutt18 can you verify if this is good to go?
Also lets see if other tests are working good, lately there have been problems with azure.
Thanks.

**Comment by kartikdutt18** [comment]
Yes it is good to go, The load save test fails because of an armadillo dependency issue and so it's not relevant to your PR. Thanks for all the work 👍.

**Comment by kartikdutt18** [comment]
Hey @ojhalakshya, Could you resolve the merge conflicts. Thanks a lot 👍

**Comment by kartikdutt18** [comment]
Also if you could do a rebase before we merge this, that would be really great. I think we have fixed every possible build so all tests should pass. Thanks again 👍.

**Comment by ojhalakshya** [comment]
> Also if you could do a rebase before we merge this, that would be really great. I think we have fixed every possible build so all tests should pass. Thanks again .

I think its up to date now after resolving the conflict.

**Comment by kartikdutt18** [comment]
Agreed, Thanks again. :)

**Comment by birm** [comment]
@kartikdutt18 do you have any other review additions or conditions other than the tests passing?

**Comment by kartikdutt18** [comment]
No, it's good to go for me 👍. Once the tests pass I think we can merge this, I think will since they did yesterday but might be better to wait for them anyway.

**Comment by ojhalakshya** [comment]
@kartikdutt18 @birm 
Thanks for the review. :)

**Comment by kartikdutt18** [comment]
Really nice to see green builds. Thanks a lot @ojhalakshya for the great work.
